### ISSUE TITLE
[BUG] Splitter unable to apply same index to both X and y

### ISSUE THREAD

**Issue by ksharma6** [sktime/sktime#7021]
**Describe the bug**
`InstanceSplitter` is unable to apply the same index to split  `X` and `y` without throwing the following error:
`ValueError: Error in InstanceSplitter.split: InstanceSplitter is a splitter of type 'instance', and requires Panel or Hierarchical time series index.`

**To Reproduce**
```python

X_train, y_train = load_arrow_head(split="train", return_type="pd-multiindex")

def gen_y_X_train_test(y,  X, cv):
        """Generate joint splits of y, X as per cv.
        """
        splitter = InstanceSplitter(cv)

        geny = splitter.split(y)
        genx = splitter.split(X)

        for (y_train, y_test), (X_train, X_test) in zip(geny, genx):
            yield y_train, y_test, X_train, X_test

xy_splits = gen_y_X_train_test(y_train, X_train, KFold(3))

for _ in xy_splits:
    print(_)
```

**Expected behavior**
I would like to see the splitter to be able to apply the same splits to both `X` and `y`. When attempt

**Additional context**
Related to: https://github.com/sktime/sktime/issues/6852#issuecomment-2272311877 and https://github.com/sktime/sktime/issues/6708#issuecomment-2250572442

**Versions**

Python dependencies:
          pip: 24.0
       sktime: 0.32.1
      sklearn: 1.4.1.post1
       skbase: 0.7.5
        numpy: 1.26.4
        scipy: 1.12.0
       pandas: 2.2.1
   matplotlib: 3.9.0
       joblib: 1.3.2
        numba: 0.59.1
  statsmodels: 0.14.2
     pmdarima: 2.0.4
statsforecast: 1.7.5
      tsfresh: 0.20.2
      tslearn: 0.6.3
        torch: None
   tensorflow: 2.16.1
tensorflow_probability: None
-->

**Comment by fkiraly** [comment]
yes, this is a more general subproblem of applying splitters to objects of different mtypes, e.g., a 1D numpy array and a `pd-multiindex`.

The problem lies in the disconnect to what `split` returns - `iloc` references to the rows of a potentially multi-indexed frame - and the `iloc` references needed for the `y` in this case, since for the `X` we apply to the level -2, whereas for the `y` we apply to the only level (the axis 0).

We need some way to (a) retrieve the integer indices for the instance, and (b) apply them both to X and y. The interface does not allow to do this easily, so the big design question is where and how to add this.

**Comment by ksharma6** [comment]
Hi @fkiraly 

I am a little stuck on how best to proceed but I thought to share some preliminary thoughts I have about potential solutions we could try:

1) Create new Splitter Class `DiffMtypeSplitter` (name has room for improvement 😅)
- Class would inherit the `BaseSplitter`
- Update `split()` param "ACCEPTED_Y_TYPES" to include multi-index mtypes
- Update `_split()` to check if `y` is a multi-index mytype and if it is, the function will return the plain indices for splitting. If not, we could use the current logic in the `_split()` for the `InstanceSplitter`

2) I think we could also do the above without creating a new class but instead just updating  `_split()` in the `InstanceSplitter` (however, I suspect there might be problems with doing it this way: namely convoluted code logic for a very specific use case)

3) We could also append a new method entirely to the `InstanceSplitter` (def split_diff_mtypes(X,y) or something of that nature)
Method could take `X` and `y` as arguments and return plain split indices for train and test sets for each.

What are  your thoughts?

**Comment by fkiraly** [comment]
I have also thought about this and found no perfect solution.

What I know we would certainly need is a function that does the following:

`apply_split(y, iloc_ix)`, which takes a `Panel` `y` and an iloc splitter output (like from sklearn `split`). Optionally, also to `Table`, or a subset relevant in `evaluate`.

Using this, we can refactor multiple places:

* the `InstanceSplitter._split`, using twice
* the new `evaluate` method

The right structure for splitters is not entirely clear, as said, but with this, at least we would unblock the `evaluate` method.

**Comment by ksharma6** [comment]
Thanks @fkiraly,

I can start with working on a helper function in `evaluate` and we can proceed from there.

**Comment by fkiraly** [comment]
Great - I think much of the `apply_split` logic can be copied from `InstanceSplitter._split`

**Comment by ksharma6** [comment]
Following up on our conversation, here is the proposed docstring for `apply_split`

```
"""Generate plain indices to split data.

    Parameters
    ----------
    y : sktime compatible tabular data container, Panel, Table scitype
         numpy1D iterable, of shape [n_instances]
    iloc_ix: generator object containing indices for splitting

    Yields
    ------
    y_train_iloc : ndarray
        The training set indices for splitting.
    y_test_iloc ndarray
        The testing set indices for splitting.
    """
```

**Comment by ksharma6** [comment]
I'm wondering if I missed something in my implementation but I am running into the same index problem
```python
def apply_split(y, x_iloc):

    if not isinstance(y, pd.MultiIndex):
        zeros = [0] * len(y)
        y = pd.MultiIndex.from_arrays([zeros, y])

    inst_ix = y.droplevel(-1).unique()
    iloc_ixer = pd.DataFrame(pd.RangeIndex(len(y)), index=y)
    
    for y_train_inst_iloc, y_test_inst_iloc in x_iloc:
        y_train_inst_loc = inst_ix[np.array(y_train_inst_iloc)]
        y_test_inst_loc = inst_ix[np.array(y_test_inst_iloc)]
        tr_np = [iloc_ixer.loc[x].to_numpy().flatten() for x in y_train_inst_loc]
        tt_np = [iloc_ixer.loc[x].to_numpy().flatten() for x in y_test_inst_loc]
        y_train_iloc = np.concatenate(tr_np)
        y_test_iloc = np.concatenate(tt_np)
        yield y_train_iloc, y_test_iloc

def gen_y_X_train_test(y, X, x_iloc):

    geny = apply_split(y, x_iloc)
    genx = apply_split(X, x_iloc)

    for (y_train, y_test), (X_train, X_test) in zip(geny, genx):
            yield y_train, y_test, X_train, X_test


```
